app/master/slave.py

```diff
@@ -244,6 +250,87 @@ def update_last_heartbeat_time(self):
     def get_last_heartbeat_time(self) -> datetime:
         return self._last_heartbeat_time
 
+    def _update_slave_registry(self):
+        """
+        Remove shutdown-ed slave from SlaveRegistry.
+        """
+        self._logger.info('Removing slave (url={}; id={}) from Slave Registry.'.format(self.url, self.id))
+        SlaveRegistry.singleton().remove_slave(slave_url=self.url)
+
+
+class SlaveRegistry(Singleton):
+    """
+    SlaveRegistry class is a singleton class which stores and maintains list of connected slaves.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.all_slaves_by_url = {}
+        self.all_slaves_by_id = {}
+        self._slave_dict_lock = Lock()
+
+    def add_slave(self, slave):
+        """
+        Add slave in SlaveRegistry.
+
+        :type slave: Slave
+        """
+        with self._slave_dict_lock:
+            self.all_slaves_by_url[slave.url] = slave
+            self.all_slaves_by_id[slave.id] = slave
+
+    def remove_slave(self, slave=None, slave_url=None):
```

Comment:
Having `slave` in the name is redundant, so just call it `remove()`

Reply:
I understand your point, but I think, as its a SlaveRegistry, `remove_slave` is more readable.

Reply:
There is a lot of code that is now just `slave_reg.remove_slave`. All of the items in the slave registry are slaves which is that's why I think its redundant to name it that way. If you follow this convention you should have `*_slave` virtually in every method name.